core/parachain/thread_pool.hpp

```diff
@@ -0,0 +1,260 @@
+            forwarding_func(func());
+          });
+    }
+
+    template <typename T, typename K>
+    void execute(std::pair<T, K> &&t) {
+      contextCall(std::move(t.first), std::move(t.second));
+    }
+
+    template <typename F, typename... Args>
+    void execute(F &&func, Args &&...args) {
+      contextCall(
+          [func{std::forward<F>(func)},
+           forwarding_func{bindArgs(std::forward<Args>(args)...)}]() mutable {
+            forwarding_func(func());
+          });
+    }
+
+    template <typename F>
+    void execute(F &&func) {
+      contextCall(std::forward<F>(func));
+    }
+
+   private:
+    friend struct ThreadQueueContext<ThreadPool>;
+
+    template <typename T, typename F>
+    void contextCall(ThreadQueueContext<T> &&t, F &&f) {
+      t.make_call(std::forward<F>(f));
+    }
+
+    template <typename F>
+    void contextCall(F &&f) {
+      ThreadQueueContext<std::weak_ptr<WorkersContext>>(context_)
+          .template make_call(std::forward<F>(f));
+    }
+
+    template <typename R, typename T, typename K>
+    void executeI(outcome::result<R> &&r, std::pair<T, K> &&t) {
+      if (r.has_value()) {
+        contextCall(
+            std::move(t.first),
+            [r{std::move(r.value())}, func{std::move(t.second)}]() mutable {
+              func(std::move(r));
+            });
+      }
+    }
+
+    template <typename R, typename F>
+    void executeI(outcome::result<R> &&r, F &&func) {
+      if (r.has_value()) {
+        contextCall(
+            [r{std::move(r.value())}, func{std::forward<F>(func)}]() mutable {
+              std::forward<F>(func)(std::move(r));
+            });
+      }
+    }
+
+    template <typename R, typename T, typename K, typename... Args>
+    void executeI(outcome::result<R> &&r, std::pair<T, K> &&t, Args &&...args) {
+      if (r.has_value()) {
+        contextCall(
+            std::move(t.first),
+            [func{std::move(t.second)},
+             r{std::move(r.value())},
+             forwarding_func{bindArgs(std::forward<Args>(args)...)}]() mutable {
+              forwarding_func(func(std::move(r)));
+            });
+      }
+    }
+
+    template <typename R, typename F, typename... Args>
+    void executeI(outcome::result<R> &&r, F &&func, Args &&...args) {
+      if (r.has_value()) {
+        contextCall(
+            [func{std::forward<F>(func)},
+             r{std::move(r.value())},
+             forwarding_func{bindArgs(std::forward<Args>(args)...)}]() mutable {
+              forwarding_func(std::forward<F>(func)(std::move(r)));
+            });
+      }
+    }
+
+    template <typename... Args>
+    auto bindArgs(Args &&...args) {
+      return std::bind(
+          [wptr{weak_from_this()}](auto &&...args) mutable {
+            if (auto self = wptr.lock()) self->executeI(std::move(args)...);
+          },
+          std::placeholders::_1,
+          std::move(args)...);
+    }
```

Comment:
Functions for function chaining are not related to thread pool.  
They should be simplified and moved to other header.